core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -2517,6 +2517,18 @@ public static Object item(Object object, Object index) {
     if (object instanceof List && index instanceof Number) {
       return arrayItem((List) object, ((Number) index).intValue());
     }
+    if (index instanceof Number) {
+      return structAccess(object, ((Number) index).intValue() - 1, ""); // 1 indexed
+    }
+    if (index instanceof String) {
```

Comment:
What are we trying to support with this if statement? Is it the following pattern?
```
person['1'] 
person[1]
```
Shouldn't we say that if it is a string it denotes access by field name and if it is an int it is access by field index?
Are there tests for this?

Reply:
We shouldn't rely on the exception for normal code flow, plus its costly especially if it happens per access.

Reply:
We support `person[$integer]` for synthetic struct types and  `person[$field_name]` for UDTs, we should not try to access `person['1']` (`'1'` cannot be a field name of the UDT as it's not a valid java identifier). I have simplified that part, if the index is a `String` we consider it a field name and we let `structAccess` handle it.

I have added a test in `core/src/test/java/org/apache/calcite/test/ReflectiveSchemaTest.java` to cover for the `person[$field_name]` access pattern (all the existing tests check that the query is legal, and that the right type is computed, none of them actually try to return the value at runtime).

I would have loved to add it to `testItemOp()` but I am not sure how to write a test which loads the bookstore schema (all the tests there target simple expressions, and I am not sure it's OK to use `CalciteAssert`).

Reply:
At the moment I have the same tests added to `ReflectiveSchemaTest.java` (`` and ``) and to `operator.iq` (as suggested by @JulianFeinauer in the JIRA ticket), there are pros/cons for both, but I think that finally `operator.iq` wins since `ITEM` is an operator, what's your take?

Reply:
Agree `operator.iq` sees more adequate for this.

Reply:
Glad you agree, @zabetak. `ReflectiveSchemaTest` should test the functionality of reflective schemas, not be a test that tests something else but uses reflective schemas because they are convenient. (We have long abused `JdbcTest` in that way; *mea culpa*.) `SqlOperatorBaseTest` and `operator.iq` are the appropriate places for testing operators (possibly also `SqlValidatorTest` and `SqlParserTest`.)

The litmus test is as follows: would someone else implementing the same feature put the test in the same place? (That has saved me a few times: I started to implement a feature and then noticed there was already a test, because the feature is already implemented.)